File: apps/app/src/lib/logo-overrides.ts
// Logo override system for tokens with custom/curated logos
// Static logo assets live in `apps/app/public/logos/...` and are served at `/logos/...`.

const LOGOS_ROUTE_BASE = "/logos" as const;
const XSTOCKS_ROUTE_BASE = `${LOGOS_ROUTE_BASE}/xstocks` as const;
const POPULAR_ROUTE_BASE = `${LOGOS_ROUTE_BASE}/popular` as const;
const XSTOCK_LOGO_EXTENSION = "png" as const;
const POPULAR_LOGO_EXTENSION = "svg" as const;

function getXstockLogoFilename(symbol: string): string | null {
    const trimmed = symbol.trim();
    if (!trimmed) return null;

    // Some tickers contain punctuation that can't be used in filenames (e.g. BRK.B → BRK_B).
    const normalizedForFilename = trimmed.replace(/[.\-]/g, "_");

    const candidates = trimmed === normalizedForFilename ? [trimmed] : [trimmed, normalizedForFilename];

    for (const candidate of candidates) {
        if (XSTOCK_LOGOS.has(candidate)) return candidate;

        const mappedFilename = SYMBOL_TO_FILENAME[candidate];
        if (mappedFilename && XSTOCK_LOGOS.has(mappedFilename)) return mappedFilename;

        // Case-insensitive match (but preserve the real filename casing from the set).
        for (const logo of XSTOCK_LOGOS) {
            if (logo.toLowerCase() === candidate.toLowerCase()) return logo;
        }
    }

    return null;
}

function normalizeSymbol(symbol: string): string {
    return symbol.trim().toLowerCase();
}

function getPopularLogoFilename(symbol: string): string | null {
    const normalized = normalizeSymbol(symbol);
    if (!normalized) return null;
    return POPULAR_SYMBOL_TO_FILENAME[normalized] ?? null;
}

// Set of available xStock logo symbols (filename without extension)
const XSTOCK_LOGOS = new Set([
    'AAPLx',
    'ABBVx',
    'ABTx',
    'ACNx',
    'ADBEx',
    'AMBRx',
    'AMDx',
    'AMZNx',
    'APPx',
    'ASMLx',
    'ASTSx',
    'AVGOx',
    'AXPx',
    'AZNx',
    'BACx',
    'BKNGx',
    'BLKx',
    'BLSHx',
    'BMNRx',
    'BRK_Bx',
    'BTBTx',
    'CLSKx',
    'CMCSAx',
    'COINx',
    'CORZx',
    'COSTx',
    'CRCLx',
    'CRMx',
    'CRWDx',
    'CSCOx',
    'CVXx',
    'DFDVx',
    'DHRx',
    'DUOLx',
    'EBAYx',
    'EXPEx',
    'FIGx',
    'FUFUx',
    'GLDx',
    'GLXYx',
    'GMEx',
    'GOOGLx',
    'GSx',
    'HDx',
    'HONx',
    'HOODx',
    'HSDTx',
    'HUTx',
    'IBMx',
    'IJRx',
    'INTCx',
    'IWMx',
    'JNJx',
    'JPMx',
    'KOx',
    'LINx',
    'LLYx',
    'LULUx',
    'MARAx',
    'MAx',
    'MCDx',
    'MDTx',
    'METAx',
    'MNSTx',
    'MRKx',
    'MRVLx',
    'MSFTx',
    'MSTRx',
    'MUx',
    'NFLXx',
    'NVDAx',
    'NVOx',
    'OKLOx',
    'OPENx',
    'ORCLx',
    'PANWx',
    'PEPx',
    'PFEx',
    'PGx',
    'PLTRx',
    'PLx',
    'PMx',
    'PYPLx',
    'QQQx',
    'RBLXx',
    'RIOTx',
    'RKLBx',
    'SBETx',
    'SCHFx',
    'SPCEx',
    'SPYx',
    'STRCx',
    'SUIGx',
    'TBLLx',
    'TEMx',
    'TGTx',
    'TMOx',
    'TMUSx',
    'TONXx',
    'TQQQx',
    'TRONx',
    'TSLAx',
    'TSMx',
    'Tx',
    'UBERx',
    'UNHx',
    'VSTx',
    'VTIx',
    'VTx',
    'Vx',
    'WBDx',
    'WENx',
    'WMTx',
    'WULFx',
    'XOMx',
]);

// Alternative symbol mappings for symbols that might differ from filenames
const SYMBOL_TO_FILENAME: Record<string, string> = {
    // Map any different API symbols to their filename
    IEMGx: 'IJRx', // Core MSCI EM might have different symbol
    SPGIx: 'SPYx', // S&P Global might map differently
    STRKx: 'STRCx', // Strategy PP Fixed
};

// Popular coin logo overrides (served from `/public/logos/popular`).
const POPULAR_SYMBOL_TO_FILENAME: Record<string, string> = {
    btc: "bitcoin",
    bitcoin: "bitcoin",
    eth: "ethereum",
    ethereum: "ethereum",
    sol: "solana",
    solana: "solana",
    ada: "cardano",
    cardano: "cardano",
    trx: "tron",
    tron: "tron",
    bnb: "bnb",
    usdt: "tether",
    tether: "tether",
    apt: "aptos",
    aptos: "aptos",
    hype: "hyperliquid",
    hyperliquid: "hyperliquid",
    mon: "monad",
    monad: "monad",
    okxbtc: "okxbtc",
};

/**
 * Get the logo URL for a token, using local override if available
 * @param symbol - Token symbol (e.g., "AAPLx")
 * @param fallbackLogoURI - Fallback logo URI from API
 * @returns Logo URL to use
 */
export function getTokenLogoURL(symbol: string | undefined, fallbackLogoURI: string | undefined): string | undefined {
    if (!symbol) return fallbackLogoURI;

    const filename = getXstockLogoFilename(symbol);
    if (filename) return `${XSTOCKS_ROUTE_BASE}/${filename}.${XSTOCK_LOGO_EXTENSION}`;

    const popularFilename = getPopularLogoFilename(symbol);
    if (popularFilename) return `${POPULAR_ROUTE_BASE}/${popularFilename}.${POPULAR_LOGO_EXTENSION}`;

    // Return fallback
    return fallbackLogoURI;
}

/**
 * Check if a token has a local logo override
 */
export function hasLocalLogo(symbol: string | undefined): boolean {
    if (!symbol) return false;
    return getXstockLogoFilename(symbol) !== null || getPopularLogoFilename(symbol) !== null;
}

/**
 * Clean token name by removing common suffixes/prefixes.
 *
 * - xStocks: "Nasdaq xStock" → "Nasdaq"
 * - LSTs: "Jupiter Staked SOL" → "Jupiter", "Marinade staked SOL (mSOL)" → "Marinade"
 * - Wrapped tokens: "Wrapped Ether" → "Ether", "Coinbase Wrapped BTC" → "BTC"
 * - Bridge markers: "(Wormhole)", "(Bridged)", etc.
 */
export function cleanTokenName(name: string | undefined): string {
    if (!name) return 'Unknown';
    return (
        name
            // xStocks
            .replace(/\s*xStock\s*$/i, '')
            // Bridge/origin markers in parens
            .replace(/\s*\(\s*(wormhole|bridged|wrapped|omnibridge|coinbase|ondo\s+tokenized)\s*\)\s*/gi, ' ')
            // Remove (mSOL) and similar LST markers
            .replace(/\s*\(\s*mSOL\s*\)\s*/gi, ' ')
            // Wrapped variants
            .replace(/^\s*coinbase\s+wrapped\s+/i, '')
            .replace(/^\s*wrapped\s+/i, '')
            .replace(/\s+wrapped\s+/gi, ' ')
            .replace(/\s+wrapped\s*$/i, '')
            // LST suffixes
            .replace(/\s+staked\s+sol(?=\s*(\(|$))/i, '')
            // Normalize whitespace
            .replace(/\s{2,}/g, ' ')
            .trim() || 'Unknown'
    );
}

File: apps/ios/Packages/AggrCore/Sources/AggrCore/Logos/LogoOverrides.swift
import Foundation

/// Port of `apps/app/src/lib/logo-overrides.ts`. Static assets live at `/logos/...` on the web origin;
/// the iOS app resolves them against `webOrigin` (defaults to https://aggr.watch).
public enum LogoOverrides {
  public static let defaultWebOrigin = URL(string: "https://aggr.watch")!

  static let xstockLogos: Set<String> = [
    "AAPLx","ABBVx","ABTx","ACNx","ADBEx","AMBRx","AMDx","AMZNx","APPx","ASMLx","ASTSx","AVGOx","AXPx","AZNx","BACx",
    "BKNGx","BLKx","BLSHx","BMNRx","BRK_Bx","BTBTx","CLSKx","CMCSAx","COINx","CORZx","COSTx","CRCLx","CRMx","CRWDx",
    "CSCOx","CVXx","DFDVx","DHRx","DUOLx","EBAYx","EXPEx","FIGx","FUFUx","GLDx","GLXYx","GMEx","GOOGLx","GSx","HDx",
    "HONx","HOODx","HSDTx","HUTx","IBMx","IJRx","INTCx","IWMx","JNJx","JPMx","KOx","LINx","LLYx","LULUx","MARAx","MAx",
    "MCDx","MDTx","METAx","MNSTx","MRKx","MRVLx","MSFTx","MSTRx","MUx","NFLXx","NVDAx","NVOx","OKLOx","OPENx","ORCLx",
    "PANWx","PEPx","PFEx","PGx","PLTRx","PLx","PMx","PYPLx","QQQx","RBLXx","RIOTx","RKLBx","SBETx","SCHFx","SPCEx",
    "SPYx","STRCx","SUIGx","TBLLx","TEMx","TGTx","TMOx","TMUSx","TONXx","TQQQx","TRONx","TSLAx","TSMx","Tx","UBERx",
    "UNHx","VSTx","VTIx","VTx","Vx","WBDx","WENx","WMTx","WULFx","XOMx",
  ]

  static let symbolToFilename: [String: String] = [
    "IEMGx": "IJRx", "SPGIx": "SPYx", "STRKx": "STRCx",
  ]

  static let popularSymbolToFilename: [String: String] = [
    "btc": "bitcoin", "bitcoin": "bitcoin", "wbtc": "bitcoin",
    "eth": "ethereum", "ethereum": "ethereum",
    "sol": "solana", "solana": "solana",
    "ada": "cardano", "cardano": "cardano",
    "trx": "tron", "tron": "tron",
    "bnb": "bnb",
    "usdt": "tether", "tether": "tether",
    "xaut": "xaut", "tether-gold": "xaut", "tether gold": "xaut",
    "apt": "aptos", "aptos": "aptos",
    "sui": "sui",
    "bp": "backpack",
    "jto": "jito",
    "hype": "hyperliquid", "hyperliquid": "hyperliquid",
    "mon": "monad", "monad": "monad",
    "okxbtc": "okxbtc",
    "2z": "doublezero", "doublezero": "doublezero",
    "uni": "uniswap", "uniswap": "uniswap",
  ]

  static func xstockFilename(for symbol: String) -> String? {
    let trimmed = symbol.trimmingCharacters(in: .whitespaces)
    guard !trimmed.isEmpty else { return nil }
    let normalized = trimmed.replacingOccurrences(of: ".", with: "_").replacingOccurrences(of: "-", with: "_")
    let candidates = trimmed == normalized ? [trimmed] : [trimmed, normalized]
    for candidate in candidates {
      if xstockLogos.contains(candidate) { return candidate }
      if let mapped = symbolToFilename[candidate], xstockLogos.contains(mapped) { return mapped }
      if let ci = xstockLogos.first(where: { $0.lowercased() == candidate.lowercased() }) { return ci }
    }
    return nil
  }

  static func normalizeSymbol(_ symbol: String) -> String {
    symbol.trimmingCharacters(in: .whitespaces).lowercased().split(whereSeparator: { $0.isWhitespace }).first.map(String.init) ?? ""
  }

  static func popularFilename(for symbol: String) -> String? {
    let normalized = normalizeSymbol(symbol)
    guard !normalized.isEmpty else { return nil }
    return popularSymbolToFilename[normalized]
  }

  /// Bundled asset name (without extension) if this symbol has a curated logo, else nil.
  public static func bundledAssetName(symbol: String?) -> String? {
    guard let symbol else { return nil }
    if let x = xstockFilename(for: symbol) { return "xstocks/\(x)" }
    if let p = popularFilename(for: symbol) { return "popular/\(p)" }
    return nil
  }

  /// `getTokenLogoURL`: curated override URL (on the web origin) or the API fallback.
  public static func tokenLogoURL(symbol: String?, fallback: String?, webOrigin: URL = defaultWebOrigin) -> URL? {
    if let symbol {
      if let x = xstockFilename(for: symbol) { return webOrigin.appending(path: "logos/xstocks/\(x).png") }
      if let p = popularFilename(for: symbol) { return webOrigin.appending(path: "logos/popular/\(p).svg") }
    }
    guard let fallback, let url = URL(string: fallback) else { return nil }
    return url
  }

  /// `cleanTokenName`.
  public static func cleanTokenName(_ name: String?) -> String {
    guard var s = name, !s.isEmpty else { return "Unknown" }
    func replace(_ pattern: String, _ with: String) {
      s = s.replacingOccurrences(of: pattern, with: with, options: [.regularExpression, .caseInsensitive])
    }
    replace(#"\s*xStock\s*$"#, "")
    replace(#"\s*\(\s*(wormhole|bridged|wrapped|omnibridge|coinbase|ondo\s+tokenized)\s*\)\s*"#, " ")
    replace(#"\s*\(\s*mSOL\s*\)\s*"#, " ")
    replace(#"^\s*coinbase\s+wrapped\s+"#, "")
    replace(#"^\s*wrapped\s+"#, "")
    replace(#"\s+wrapped\s+"#, " ")
    replace(#"\s+wrapped\s*$"#, "")
    replace(#"\s+staked\s+sol(?=\s*(\(|$))"#, "")
    replace(#"\s{2,}"#, " ")
    let out = s.trimmingCharacters(in: .whitespaces)
    return out.isEmpty ? "Unknown" : out
  }
}
